File: solutions/python/yacht/1/yacht.py
# Score categories.
# Change the values as you see fit.
YACHT = 'YACHT'
ONES = 'ONES'
TWOS = 'TWOS'
THREES = 'THREES'
FOURS = 'FOURS'
FIVES = 'FIVES'
SIXES = 'SIXES'
FULL_HOUSE = 'FULL_HOUSE'
FOUR_OF_A_KIND = 'FOUR_OF_A_KIND'
LITTLE_STRAIGHT = 'LITTLE_STRAIGHT'
BIG_STRAIGHT = 'BIG_STRAIGHT'
CHOICE = 'CHOICE'


def score(dice, category):
    dice.sort()

    if category == 'YACHT' and sum(dice) == dice[0] * 5:
        return 50
    if category == 'BIG_STRAIGHT' and dice == [2,3,4,5,6]:
        return 30
    if category == 'LITTLE_STRAIGHT' and dice == [1,2,3,4,5]:
        return 30

    counts = dict((i, dice.count(i)) for i in dice)
    threes_sum = None
    twos_sum = None
    fours_sum = None

    for pips in counts:
        match counts.get(pips):
            case 4:
                fours_sum = pips * 4
            case 3:
                threes_sum = pips * 3
            case 2:
                twos_sum = pips * 2
 
    if category == 'FULL_HOUSE' and threes_sum != None and twos_sum != None:
            return threes_sum + twos_sum
    if category == 'FOUR_OF_A_KIND' and fours_sum != None:
            return fours_sum
    if category == 'FOUR_OF_A_KIND' and sum(dice) == dice[0] * 5:
            return dice[0] * 4
    if category == 'SIXES': return 6 * dice.count(6)
    if category == 'FIVES': return 5 * dice.count(5)
    if category == 'FOURS': return 4 * dice.count(4)
    if category == 'THREES': return 3 * dice.count(3)
    if category == 'TWOS': return 2 * dice.count(2)
    if category == 'ONES': return 1 * dice.count(1)
    if category == 'CHOICE': return sum(dice)

    return 0

File: solutions/cpp/yacht/1/yacht.cpp
#include "yacht.h"

namespace yacht {

// TODO: add your solution here
    int score(vector<int> input, string category) {
        if (category == "yacht") {
            if (input[0] == input [1] &&
                input[1] == input [2] &&
                input[2] == input [3] &&
                input[3] == input [4]) {
                    return 50;
                }
        }
        if (category == "choice") {
            int sum{};
            for (int i : input) {
                sum += i;
            }
            return sum;
        }
        if (category == "big straight") {
            int sum{};
            for (int i : input) {
                if (i != 1) {
                    int count = std::count(input.begin(), input.end(), i);

                    if (count == 1) {
                        sum += i;
                    }
                }
            }

            if (sum == 20) {
                return 30;
            }
        }
        if (category == "little straight") {
            int sum{};
            for (int i : input) {
                if (i != 6) {
                    int count = std::count(input.begin(), input.end(), i);

                    if (count == 1) {
                        sum += i;
                    }
                }
            }

            if (sum == 15) {
                return 30;
            }
        }
        if (category == "four of a kind") {
            for (int i : input) {
                int count = std::count(input.begin(), input.end(), i);

                if (count >= 4) {
                    return 4 * i;
                }
            }
        }
        if (category == "full house") {
            int sum{};
            bool valid = true;

            for (int i : input) {
                int count = std::count(input.begin(), input.end(), i);

                if (count == 2 || count == 3) {
                    sum += i;
                } else {
                    valid = false;
                    break;
                }
            }

            if (valid == true) {
                return sum;
            }
        }
        if (category == "sixes") {
            int count{};
            
            for (int i :input) {
                if (i == 6) {
                    count += 1;
                }
            }
            return count * 6;
        }
        if (category == "fives") {
         int count{};
            
            for (int i :input) {
                if (i == 5) {
                    count += 1;
                }
            }
            return count * 5;
        }
        if (category == "fours") {
        int count{};
            
            for (int i :input) {
                if (i == 4) {
                    count += 1;
                }
            }
            return count * 4;
        }
        if (category == "threes") {
        int count{};
            
            for (int i :input) {
                if (i == 3) {
                    count += 1;
                }
            }
            return count * 3;
        }
        if (category == "twos") {
            int count{};
            
            for (int i :input) {
                if (i == 2) {
                    count += 1;
                }
            }
            return count * 2;
        }
        if (category == "ones") {
        int count{};
            
            for (int i :input) {
                if (i == 1) {
                    count += 1;
                }
            }
            return count;
        }
        return 0;
    }
}  // namespace yacht
